### PR TITLE
Describe wasi-filesystem's path sandboxing behavior

### PR DESCRIPTION
**Pull request by sunfishcode**
Add a new path-resolution.md document describing CloudABI-style path resolution behavior. wit/types.wit previously had a brief description of this behavior, however this new document presents it in much more detail, with a description of the sandboxing properties and a discussion of implementation techniques.

### CODE REVIEW

File: path-resolution.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+by the base handle. Information about the filesystem outside of the base
+directory handles is not visible. In particular, it's not permitted to use
+paths that temporarily step outside the sandbox with something like
+"../../../stuff/here", even if the final resolved path is back inside the
+sandbox, because that would leak information about the existence of
+directories outside the sandbox.
+
+And, the sandboxing is designed to be implementable even in the presence
+of outside processes accessing the same filesystem, including renaming,
+unlinking, and creating new files and directories.
```
Comment (guybedford) [inline path-resolution.md:29]:
```suggestion
Importantly, the sandboxing is designed to be implementable even in the presence
of outside processes accessing the same filesystem, including renaming,
unlinking, and creating new files and directories.
```

File: path-resolution.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+created or renamed by other entities with access to the filesystem.
+
+## Implementation notes
+
+On Linux, `openat2` with `RESOLVE_BENEATH` may be used to implement many
+system calls other than just "open" by utilizing Linux's`O_PATH` and
+"/proc/self/fd" features.
+
+On Windows, the [`NtCreateFile`] function can accept a directory handle and
+can behave like an `openat` function.
```
Comment (guybedford) [inline path-resolution.md:73]: I take it these are then optimizations to what has been described above? Perhaps framing it that way would help?

**Review by guybedford — APPROVED** [review]

File: path-resolution.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+paths that temporarily step outside the sandbox with something like
+"../../../stuff/here", even if the final resolved path is back inside the
+sandbox, because that would leak information about the existence of
+directories outside the sandbox.
+
+And, the sandboxing is designed to be implementable even in the presence
+of outside processes accessing the same filesystem, including renaming,
+unlinking, and creating new files and directories.
+
+## Implementing path resolution manually
```
[the remark this answers is not in the document]
Comment (sunfishcode) [reply path-resolution.md]: The symlink one describes user-visible symlink behavior. But I've now moved the "Implementing path resolution manually" and "Implementation notes" into a `Host Implementation` section.

File: path-resolution.md (hunk 2 above)
Comment (sunfishcode) [reply path-resolution.md]: The Linux one is an optimization, and I've now added words mentioning that. The Windows one provides an `openat`-like primitive useful for implementing the manual algorithm, and I've now added words describing that too.

**Comment by sunfishcode** [comment]
I've answered all the questions above; I'm going to go ahead and merge this now. If there are any further questions; please file new issues!

**Pull request merged**